File: pwm/python/pwm_sysfs.py
import sys


class PwmSysfs:
    def __init__(self, chip_no: int, channel_no: int, period_ns: int, duty_cycle_ns: int):
        self._pwm_sysfs_base = "/sys/class/pwm"
        self._is_initialized = False

        self.m_chip_no = chip_no
        self.m_channel_no = channel_no
        self.m_period_ns = period_ns
        self.m_duty_cycle_ns = duty_cycle_ns

    def __del__(self):
        if self._is_initialized:
            self.set_enable("0")

    def initialize(self) -> int:
        try:
            with open(f"{self._pwm_sysfs_base}/pwmchip{self.m_chip_no}/export", "w") as export_file:
                export_file.write(str(self.m_channel_no))
        except OSError:
            print("Note: PWM channel might already be exported", file=sys.stderr)

        try:
            with open(
                f"{self._pwm_sysfs_base}/pwmchip{self.m_chip_no}/pwm{self.m_channel_no}/period", "w"
            ) as period_file:
                period_file.write(str(self.m_period_ns))
        except OSError as err:
            print(err, file=sys.stderr)
            print("Failed to set PWM period", file=sys.stderr)
            return 1

        try:
            with open(
                f"{self._pwm_sysfs_base}/pwmchip{self.m_chip_no}/pwm{self.m_channel_no}/duty_cycle", "w"
            ) as duty_cycle_file:
                duty_cycle_file.write(str(self.m_duty_cycle_ns))
        except OSError as err:
            print(err, file=sys.stderr)
            print("Failed to set PWM duty cycle", file=sys.stderr)
            return 1

        self._is_initialized = True
        return 0

    def is_initialized(self) -> bool:
        return self._is_initialized

    def set_enable(self, value: str) -> int:
        try:
            with open(
                f"{self._pwm_sysfs_base}/pwmchip{self.m_chip_no}/pwm{self.m_channel_no}/enable", "w"
            ) as enable_file:
                enable_file.write(value)
        except OSError:
            return 1

        return 0

File: pwm/cpp/pwm_sysfs.cpp
// Copyright (c) 2025 by T3 Foundation. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//     https://www.t3gemstone.org/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

#include "pwm_sysfs.h"
#include <cstdio>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>

PwmSysfs::PwmSysfs(uint8_t chip_no, uint8_t channel_no, uint32_t period_ns, uint32_t duty_cycle_ns)
    : m_chip_no {chip_no}, m_channel_no {channel_no}, m_period_ns {period_ns}, m_duty_cycle_ns {duty_cycle_ns} {};

PwmSysfs::~PwmSysfs()
{
    if (m_is_initialized)
    {
        set_enable("0");
    }
};

int PwmSysfs::initialize()
{
    snprintf(m_path, sizeof(m_path), "%s/pwmchip%d/export", m_pwm_sysfs.data(), m_chip_no);
    snprintf(m_value, sizeof(m_value), "%d", m_channel_no);
    if (write_to_file(m_path, m_value) < 0)
    {
        printf("Note: PWM channel might already be exported\n");
    }

    snprintf(m_path, sizeof(m_path), "%s/pwmchip%d/pwm%d/period", m_pwm_sysfs.data(), m_chip_no, m_channel_no);
    snprintf(m_value, sizeof(m_value), "%d", m_period_ns);
    if (write_to_file(m_path, m_value) < 0)
    {
        fprintf(stderr, "Failed to set PWM period\n");
        return 1;
    }

    snprintf(m_path, sizeof(m_path), "%s/pwmchip%d/pwm%d/duty_cycle", m_pwm_sysfs.data(), m_chip_no, m_channel_no);
    snprintf(m_value, sizeof(m_value), "%d", m_duty_cycle_ns);
    if (write_to_file(m_path, m_value) < 0)
    {
        fprintf(stderr, "Failed to set PWM duty cycle\n");
        return 1;
    }

    m_is_initialized = true;
    return 0;
};

int PwmSysfs::is_initialized()
{
    return m_is_initialized;
}

int PwmSysfs::set_enable(std::string_view value)
{
    snprintf(m_path, sizeof(m_path), "%s/pwmchip%d/pwm%d/enable", m_pwm_sysfs.data(), m_chip_no, m_channel_no);
    if (write_to_file(m_path, value) < 0)
    {
        return 1;
    }
    return 0;
};

int PwmSysfs::write_to_file(std::string_view path, std::string_view value)
{
    int fd = open(path.data(), O_WRONLY);
    if (fd < 0)
    {
        perror("open");
        return -1;
    }

    if (write(fd, value.data(), strlen(value.data())) < 0)
    {
        perror("write");
        close(fd);
        return -1;
    }

    close(fd);
    return 0;
};
